Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -1186,15 +1189,34 @@ namespace Babylon
                 env.Global().Set(JS_CLASS_NAME, func);
             }
 
-            static Napi::Promise CreateAsync(const Napi::CallbackInfo& info)
+            static Napi::Promise CreateAsync(const Napi::CallbackInfo& info, const JsRuntimeScheduler& runtimeScheduler)
             {
-                auto jsSession = info.Env().Global().Get(JS_CLASS_NAME).As<Napi::Function>().New({info[0]});
-                auto& session = *XRSession::Unwrap(jsSession);
-
-                session.m_xr.BeginSession(info.Env());
+                auto jsSession = Napi::Persistent(info.Env().Global().Get(JS_CLASS_NAME).As<Napi::Function>().New({info[0]}));
+                auto& session = *XRSession::Unwrap(jsSession.Value());
 
                 auto deferred = Napi::Promise::Deferred::New(info.Env());
-                deferred.Resolve(jsSession);
+                session.m_xr.BeginSession(info.Env(), runtimeScheduler)
+                    .then(runtimeScheduler,
+                        arcana::cancellation::none(),
+                        [deferred, jsSession = std::move(jsSession), env = info.Env()](const arcana::expected<void, std::exception_ptr>& result) {
+                            if (result.has_error())
+                            {
+                                try
+                                {
+                                    std::rethrow_exception(result.error());
+                                }
+                                catch (const std::exception& e)
+                                {
+                                    deferred.Reject(Napi::Error::New(env, e.what()).Value());
+                                    throw e;
```

Comment:
We shouldn't re-throw this should we? The promise has been rejected, so the JS side should be able to handle the exception. If we re-throw here, it doesn't go anywhere that can be handled, right?

Reply:
You're right, removed.